Introduction of data streaming in adapters and fetch.js

MichalCz opened the pull request:
Fixes #452 .

The change adds the option to expose a streamed version of the adapter data via the new `fetchStream` method. It also allows handling of that fetched data in fetch.js.

In streamed mode, the data is parsed and saved into the database already during downloading of the source, which greatly reduced memory consumption, since data is never fully aggregated into arrays kept in memory.

If the adapted does not support streaming (does not export the new method), the processed data is turned into stream and at least the fetch.js part of processing is not consuming extra memory.

jflasher commented:
Thanks @MichalCz! Will try to check this out soon.

jflasher commented on .editorconfig:
```diff
@@ -0,0 +1,10 @@
+[*]
```
Please delete this file as I think it's editor specific?

MichalCz replied:
Will do

jflasher commented on adapters/eea-direct.js:
```diff
@@ -2,146 +2,104 @@ import { acceptableParameters, convertUnits } from '../lib/utils';
+          return stream;
+        })
+        .CSVParse({header: true, delimiter: ',', skipEmptyLines: true})
+        // TODO: it would be good to provide the actual last fetch time so that we can filter already inserted items in a better way
```
Good point! I believe we're actually storing this in Redis for another purpose so we may be able to get this here without too much difficulty, but I think this is an ok thing to leave undone for now.

jflasher commented on adapters/eea-direct.js:
```diff
@@ -2,146 +2,104 @@ import { acceptableParameters, convertUnits } from '../lib/utils';
 import { REQUEST_TIMEOUT } from '../lib/constants';
```
Ok, so I tested this out and had mixed results. The first time I was running it on Slovakia which is a smallish source (~200 matching measurements), it ran fine. I compared to data coming from the current process and it matched, so that was all great. However, coming back to Slovakia a bit later, after new data was being provided, I ran into the errors below. Note that I also saw these errors when trying to do the larger sources like `EEA Germany` and `EEA France`. Was able to run and test just the specific source with `node index.js -d -s 'EEA Slovakia'`. This looks like perhaps the stream is losing data during it's processing and it doesn't like that?

```
(node:5994) UnhandledPromiseRejectionWarning: Error: Too few fields: expected 24 fields but parsed 3
    at Object.chunk (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet/lib/string-stream.js:50:60)
    at ReadableStreamStreamer.ChunkStreamer.parseChunk (/Users/jflasher/workspace/openaq-fetch/node_modules/papaparse/papaparse.js:488:18)
    at ReadableStreamStreamer.<anonymous> (/Users/jflasher/workspace/openaq-fetch/node_modules/papaparse/papaparse.js:789:11)
    at ReadableStreamStreamer._streamData (/Users/jflasher/workspace/openaq-fetch/node_modules/papaparse/papaparse.js:1585:25)
    at ReadableStreamStreamer.<anonymous> (/Users/jflasher/workspace/openaq-fetch/node_modules/papaparse/papaparse.js:808:9)
    at StringStream.<anonymous> (/Users/jflasher/workspace/openaq-fetch/node_modules/papaparse/papaparse.js:1585:25)
    at StringStream.emit (events.js:165:20)
    at endReadableNT (_stream_readable.js:1101:12)
    at process._tickCallback (internal/process/next_tick.js:152:19)
(node:5994) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 16760)
(node:5994) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:5994) UnhandledPromiseRejectionWarning: Error: Too few fields: expected 24 fields but parsed 5
    at Object.chunk (/Users/jflasher/workspace/openaq-fetch/node_modules/scramjet/lib/string-stream.js:50:60)
    at ReadableStreamStreamer.ChunkStreamer.parseChunk (/Users/jflasher/workspace/openaq-fetch/node_modules/papaparse/papaparse.js:488:18)
    at ReadableStreamStreamer.<anonymous> (/Users/jflasher/workspace/openaq-fetch/node_modules/papaparse/papaparse.js:789:11)
    at ReadableStreamStreamer._streamData (/Users/jflasher/workspace/openaq-fetch/node_modules/papaparse/papaparse.js:1585:25)
    at ReadableStreamStreamer.<anonymous> (/Users/jflasher/workspace/openaq-fetch/node_modules/papaparse/papaparse.js:808:9)
    at StringStream.<anonymous> (/Users/jflasher/workspace/openaq-fetch/node_modules/papaparse/papaparse.js:1585:25)
    at StringStream.emit (events.js:165:20)
    at endReadableNT (_stream_readable.js:1101:12)
    at process._tickCallback (internal/process/next_tick.js:152:19)
(node:5994) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 17643)
```

MichalCz replied:
It seems that some lines are not full inn that data. In the previous implementation such an entry would be simply omitted - I'll fix it

jflasher replied:
Ha yeah, not surprising. Source data across various adapters often has issues.

jflasher commented on fetch.js:
```diff
@@ -380,6 +377,7 @@ var runTasks = function () {
     let timeEnded = new Date();
     if (err) {
       log.error('Error during fetching of data sources.');
+      log.error(err.stack);
```
Please remove this or change to a `console.error()`. The logging goes to Papertrail and we've had issues with large stacks trying to get sent to the service which will crash the process. Worst case if it's changed to `console.error()` we can log in and check the logs from the process locally.

MichalCz replied:
Ok

jflasher commented on adapters/airnow.js:
```diff
@@ -16,149 +19,116 @@ exports.fetchData = function (source, cb) {
-    const file = `HourlyData/${dateString}.dat`;
-    const lineToObj = function (line) {
+  getStream('Locations/monitoring_site_locations.dat')
+    .map(line => {
```
Getting a `getStream(...).map is not a function` error for this adapter here.

MichalCz replied:
Will check

jflasher commented on fetch.js:
```diff
@@ -20,7 +20,10 @@ var argv = require('yargs')
   .alias('h', 'help')
```
Putting this here as I'm not sure where it's actually coming from. When doing a full test like `node index.js`, these errors start to pop up, but not when testing individual sources. We can obviously disable the limit, but this probably points to something not great happening.

```
(node:5540) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added. Use emitter.setMaxListeners() to increase limit
```

MichalCz replied:
Not really. That's just a message indicating more than 11 muxed stream - its safe to ignore, but I'll check it out.

MichalCz replied:
I published a new version of scramjet that adds control to the listener number. I'll update the package.json here tomorrow and we'll see if it solved all of those.

MichalCz commented:
I'll check these issues and fix them ETA next week.

jflasher commented:
Thanks!

MichalCz commented:
I fixed the minor things and removed the sample code in airnow.js. The unhandled promise rejection and eventemitter warnings are still to do.

@jflasher can you check the changes I made now?

jflasher commented:
@MichalCz Thanks and the other changes look good.

jflasher commented:
@MichalCz thanks for this! I pulled this in and removed/installed scramjet to ensure I'm on the new version, but looks like I'm stilling see `MaxListenersExceededWarning`s, though maybe less than before? Should this automatically be getting handled by scramjet or is there a flag I need to enable?

Also, I saw 

```
internal/streams/legacy.js:59
      throw er; // Unhandled stream error in pipe.

Error: ESOCKETTIMEDOUT
    at ClientRequest.<anonymous> (/Users/jflasher/workspace/openaq-fetch/node_modules/request/request.js:813:19)
    at Object.onceWrapper (events.js:255:19)
    at ClientRequest.emit (events.js:160:13)
    at Socket.emitTimeout (_http_client.js:702:34)
    at Object.onceWrapper (events.js:255:19)
    at Socket.emit (events.js:160:13)
    at Socket._onTimeout (net.js:415:8)
    at ontimeout (timers.js:458:11)
    at tryOnTimeout (timers.js:296:5)
    at Timer.listOnTimeout (timers.js:259:5)
```

Thanks for all the help here!

MichalCz commented:
No. It should be automatically handled. I'd need to be able to run the full import to reproduce - can you give me some instructions how to do that?

jflasher commented:
I'd recommend not running with `--dryrun` as the error messages will get lost in the output of the data. So this just means you'll need a PostgreSQL database to let the process write to (which you may have already set up). Then you can just run `node index.js` and it'll run the whole ingest. You may lose one or two sources due to missing API keys, but EEA is the big one and doesn't need a key so that shouldn't be a problem. Thanks!

MichalCz commented:
I made some changes and released a new version of the framework - now there's only the columns issue to be fixed. I'll look into that in the coming days.

Also resolved some conflicts.

MichalCz commented:
Joe - I'm now running a full import so please hold until I confirm that I haven't run into additional issues.

The requested fix on EEA Direct ended up in manually generating objects from csv headers. To appears to be all good on a full DB run on EEA France.

MichalCz commented:
Should be all good now. Let me know if you see any more issues.

jflasher commented:
Thanks for all the work @MichalCz made a few small changes so this will be superseded by https://github.com/openaq/openaq-fetch/pull/495

Will try to run on staging which has a pretty undersized instance, so we'll see if it runs there. Trying to merge this with some time to monitor before I head on vacation at the end of the week.

MichalCz commented:
Cool. :)

I'm happy to help out. :)
